cpp/src/arrow/compute/kernels/scalar_cast_test.cc

```diff
@@ -2765,6 +2765,26 @@ TEST(Cast, ExtensionTypeToIntDowncast) {
   }
 }
 
+TEST(Cast, PrimitiveToExtension) {
+  auto primitive_array = ArrayFromJSON(uint8(), "[0, 1, 3]");
+  auto extension_array = SmallintArrayFromJSON("[0, 1, 3]");
+  CastOptions options;
+  options.to_type = smallint();
+  CheckCast(primitive_array, extension_array, options);
+}
```

Comment:
Also add a test similar to `ExtensionTypeToIntDowncast`, but downcasting from primitive to extension?
(to check forwarding of cast options)